src/privatesend/privatesend-util.h

```diff
@@ -34,4 +36,96 @@ class CKeyHolderStorage
+    bool fKeepKeys{false};
+    // Protect vecOutputs
+    mutable CCriticalSection cs_outputs;
+    // Contains all outputs already added to the transaction
+    std::vector<std::unique_ptr<CTransactionBuilderOutput>> vecOutputs;
+    // Needed by CTransactionBuilderOutput::UpdateAmount to lock cs_outputs
+    friend class CTransactionBuilderOutput;
+
+public:
+    CTransactionBuilder(CWallet* pwalletIn, const CompactTallyItem& tallyItemIn);
+    ~CTransactionBuilder();
+    // Check it would be possible to add a single output with the amount nAmount. Returns true if its possible and false if not.
+    bool CouldAddOutput(CAmount nAmountOutput) const;
+    // Check if its possible to add multiple outputs as vector of amounts. Returns true if its possible to add all of them and false if not.
+    bool CouldAddOutputs(const std::vector<CAmount>& vecOutputAmounts) const;
+    // Add an output with the amount nAmount. Returns a pointer to the output if it could be added and nullptr if not due to insufficient amount left.
+    CTransactionBuilderOutput* AddOutput(CAmount nAmountOutput = 0);
+    // Get amount we had available when we started
+    CAmount GetAmountInitial() const { return tallyItem.nAmount; }
+    // Get the amount currently left to add more outputs. Does respect fees.
+    CAmount GetAmountLeft() const { return GetAmountInitial() - GetAmountUsed() - GetFee(GetBytesTotal()); }
+    // Check if an amounts should be considered as dust
+    bool IsDust(CAmount nAmount) const;
+    // Get the total number of added outputs
+    int CountOutputs() const { return vecOutputs.size(); }
+    // Create and Commit the transaction to the wallet
+    bool Commit(std::string& strResult);
+    // Convert to a string
+    std::string ToString() const;
+
+private:
+    // Clear the output vector and keep/return the included keys depending on the value of fKeepKeys
+    void Clear();
+    // Get the total number of bytes used already by this transaction
+    unsigned int GetBytesTotal() const;
+    // Helper to calculate static amount left by simply subtracting an used amount and a fee from a provided initial amount.
+    static CAmount GetAmountLeft(CAmount nAmountInitial, CAmount nAmountUsed, CAmount nFee);
+    // Get the amount currently used by added outputs. Does not include fees.
+    CAmount GetAmountUsed() const;
+    // Get fees based on the number of bytes and the feerate set in CoinControl.
+    // NOTE: To get the total transaction fee this should only be called once with the total number of bytes for the transaction to avoid
+    // calling CFeeRate::GetFee multiple times with subtotals as this may add rounding errors with each further call.
+    CAmount GetFee(unsigned int nBytes) const;
+    // Helper to get GetSizeOfCompactSizeDiff(vecOutputs.size(), vecOutputs.size() + nAdd)
+    int GetSizeOfCompactSizeDiff(size_t nAdd) const;
```

Comment:
kindof nit: can you make these comments doxygen? `///`?

Reply:
done in c30df0d